### PR TITLE
[GH-835] Randomize Zone shrinking position

### LINKED ISSUE #835: [FEATURE] - Randomize zone's center position
The position randomizing should have these three constraints:

- Zone should be bigger, so its start outside the map
- Zone center position does not move around the map
- Zone center position can’t be anywhere in the map, we have to define a zone where it can appear

### PR DESCRIPTION
**Pull request by tvillegas98**
> [!WARNING]  
> Merged alonside this [Front PR](https://github.com/lambdaclass/champions_of_mirra/pull/2028)

## Motivation

In search of improving gameplay and making each match different, we want the zone to have a random center position. In our current state, the zone always shrinks in the center of the map

Closes #835 

## Summary of changes

- Added a shrinking position for the zone
- Added a `zone_random_position_radius` to set a smaller radius / bigger radius for the zone to get a random shrinking center position
- Now we send the shrinking center position to the front
 
DISCLAIMER: Although this PR introduces logic to randomize the zone shrinking position, we're not gonna use it because in the client it seems to be a bit complex to fix the zone vfx. I have achieved this disable state by setting the `zone_random_position_radius` field set to 0.

image

## How to test it?

- Run again the seeds or modify the new added fields in the configurator.
- Play a match and see that the zone shrinks in a different position

## Checklist
- [x] Tested the changes locally.
- [x] Reviewed the changes on GitHub, line by line.
- [ ] This change requires new documentation.
  - [ ] Documentation has been added/updated.

### CODE REVIEW

File: apps/configurator/lib/configurator_web/controllers/game_configuration_html/show.html.heex [hunk 1]
```diff
@@ -21,7 +21,9 @@
   <:item title="Zone stop interval ms"><%= @game_configuration.zone_stop_interval_ms %></:item>
   <:item title="Zone start interval ms"><%= @game_configuration.zone_start_interval_ms %></:item>
   <:item title="Zone damage interval ms"><%= @game_configuration.zone_damage_interval_ms %></:item>
-  <:item title="Zone damage"><%= @game_configuration.zone_damage %></:item>
+  <:item title="Zone damage"><%= @game_configuration.zone_start_radius %></:item>
```
Comment (agustinesco) [inline apps/configurator/lib/configurator_web/controllers/game_configuration_html/show.html.heex:24]: I think this one is wrong right?

File: apps/configurator/lib/configurator_web/controllers/game_configuration_html/show.html.heex [hunk 2]
```diff
@@ -21,7 +21,9 @@
...
+  <:item title="Zone map radius"><%= @game_configuration.zone_random_position_radius %></:item>
+  <:item title="Zone random position radius"><%= @game_configuration.zone_damage %></:item>
```
Comment (agustinesco) [inline apps/configurator/lib/configurator_web/controllers/game_configuration_html/show.html.heex:26]: This one too!

**Comment by agustinesco** [comment]
https://github.com/user-attachments/assets/2c5c7539-daac-477a-9dae-1198ba25b746

Even tho the zone didn't start to shrink i receive damage from it, maybe we should do something with it to avoid this situation and still have the random shrink point

**Comment by tvillegas98** [comment]
> Even tho the zone didn't start to shrink i receive damage from it, maybe we should do something with it to avoid this situation and still have the random shrink point

That's actually happening in main too 🤔 .

File: apps/configurator/lib/configurator_web/controllers/game_configuration_html/show.html.heex (hunk 1 above)
Comment (tvillegas98) [reply apps/configurator/lib/configurator_web/controllers/game_configuration_html/show.html.heex:24]: Whoops

File: apps/configurator/lib/configurator_web/controllers/game_configuration_html/show.html.heex (hunk 2 above)
Comment (tvillegas98) [reply apps/configurator/lib/configurator_web/controllers/game_configuration_html/show.html.heex]: Fixed!

File: apps/configurator/lib/configurator_web/controllers/game_configuration_html/show.html.heex [hunk 3]
```diff
@@ -21,7 +21,8 @@
   <:item title="Zone stop interval ms"><%= @game_configuration.zone_stop_interval_ms %></:item>
   <:item title="Zone start interval ms"><%= @game_configuration.zone_start_interval_ms %></:item>
   <:item title="Zone damage interval ms"><%= @game_configuration.zone_damage_interval_ms %></:item>
-  <:item title="Zone damage"><%= @game_configuration.zone_damage %></:item>
```
Comment (agustinesco) [inline apps/configurator/lib/configurator_web/controllers/game_configuration_html/show.html.heex:24]: We lost this

Comment (tvillegas98) [reply apps/configurator/lib/configurator_web/controllers/game_configuration_html/show.html.heex:24]: Whoops

Comment (tvillegas98) [reply apps/configurator/lib/configurator_web/controllers/game_configuration_html/show.html.heex:24]: Done in [fix: return zone damage](https://github.com/lambdaclass/mirra_backend/pull/849/commits/18eee02271a9e7c6ebace70c57f3517fcaa512f2)

**Review by agustinesco — APPROVED** [review]
LGTM!

File: apps/arena/lib/arena/game_updater.ex [hunk 4]
```diff
@@ -824,12 +824,22 @@ defmodule Arena.GameUpdater do
     {bushes, last_id} = initialize_bushes(config.map.bushes, last_id)
     {pools, last_id} = initialize_pools(config.map.pools, last_id)
 
+    zone_shrinking_center_position =
+      random_position_in_map(
+        0.0,
```
Comment (Nico-Sanchez) [inline apps/arena/lib/arena/game_updater.ex:829]: What's this zero hardcoded value?

**Review by Nico-Sanchez — CHANGES_REQUESTED** [review]
Tested in Unity with client branch and found the following issues:
1. Zone looks weird behind below rocks and I can walk in it's border without taking any damage.

<video src="https://github.com/user-attachments/assets/602da93a-df91-404e-8a80-9d8b97fbabd4"></video>

2. Position is not random, It always end in the center. Also the circle is too sharp-edged.

<video src="https://github.com/user-attachments/assets/6bb9e8f4-08e9-48bc-b939-2b106a7284d4">

File: apps/serialization/messages.proto [hunk 5]
```diff
@@ -365,6 +365,7 @@ message Zone {
   bool enabled = 2;
   int64 next_zone_change_timestamp = 3;
   bool shrinking = 4;
+  Position shrinking_center_position = 5;
```
[the remark this answers is not in the document]
Comment (tvillegas98) [reply apps/serialization/messages.proto:368]: I agree

**Comment by tvillegas98** [comment]
> Also the circle is too sharp-edged

It was a bug in the front, nice catch!

File: apps/arena/lib/arena/game_updater.ex (hunk 4 above)
Comment (tvillegas98) [reply apps/arena/lib/arena/game_updater.ex:829]: That function has weird parameters but that `0.0` represents the center of the circle that's going to be created to select a random position within that circle.

**Comment by Nico-Sanchez** [comment]
We won't merge this until we fix the zone initial position and behavior.

**Comment by tvillegas98** [comment]
> We won't merge this until we fix the zone initial position and behavior.

Fixed in [feat: enable the zone only when start timer has finished](https://github.com/lambdaclass/mirra_backend/pull/849/commits/cfa3f0d71e665bf00e988513d483adb5eed8a502) ! let me know if you find something else

**Comment by Nico-Sanchez** [comment]
I'll close this until we define what to do with this feature.

**Pull request closed without merging**